### PR TITLE
Port Apply to Dotty

### PR DESCRIPTION
**Pull request by andrzejressel**
Same as https://github.com/typelevel/kittens/pull/439 - requires 3.1.2 for tuples.

### CODE REVIEW

File: core/src/test/scala-3/cats/derived/ApplySuite.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  import TestDefns._
+
+  def testApply(context: String)(using
+      caseClassWOption: Apply[CaseClassWOption],
+      optList: Apply[OptList],
+      // Requires scala 3.1.2
+//      andInt: Apply[AndInt],
+      interleaved: Apply[Interleaved],
+      listBox: Apply[ListBox]
+  ): Unit = {
```
Comment (joroKr21) [inline core/src/test/scala-3/cats/derived/ApplySuite.scala:34]: We should make this function `inline` to get rid the `using` clauses - see `FunctorSuite` for an example.

**Review by joroKr21 — COMMENTED** [review]
Nice! 👍 Thank you for working on this!
Some minor comments, but otherwise LGTM

**Comment by andrzejressel** [comment]
So I've tried to combine applyTest with getting isomorphism
```
diff --git a/core/src/test/scala-3/cats/derived/ApplySuite.scala b/core/src/test/scala-3/cats/derived/ApplySuite.scala
index 5150e71..4f081e8 100644
--- a/core/src/test/scala-3/cats/derived/ApplySuite.scala
+++ b/core/src/test/scala-3/cats/derived/ApplySuite.scala
@@ -20,26 +20,25 @@ import cats.Apply
 import cats.laws.discipline.{ApplyTests, SerializableTests}
 import cats.laws.discipline.SemigroupalTests.Isomorphisms
 import scala.compiletime.*
+import org.typelevel.discipline.Laws
+import org.scalacheck.Arbitrary
 
 class ApplySuite extends KittensSuite {
   import ApplySuite._
   import TestDefns._
 
-  inline def applyTests[F[_]]: ApplyTests[F] =
-    ApplyTests[F](summonInline)
+  inline def applyTests[F[_]]: Laws#RuleSet = {
+    given apply: Apply[F] = summonInline[Apply[F]]
+    given iso: Isomorphisms[F] = Isomorphisms.invariant(apply)
+    ApplyTests[F](apply).apply[Int, String, Long]
+  }
 
   inline def testApply(inline context: String): Unit = {
-    given Isomorphisms[CaseClassWOption] = Isomorphisms.invariant(summonInline[Apply[CaseClassWOption]])
-    given Isomorphisms[OptList] = Isomorphisms.invariant(summonInline[Apply[OptList]])
-    //    given isoAndInt: Isomorphisms[AndInt] = Isomorphisms.invariant(andInt)
-    given Isomorphisms[Interleaved] = Isomorphisms.invariant(summonInline[Apply[Interleaved]])
-    given Isomorphisms[ListBox] = Isomorphisms.invariant(summonInline[Apply[ListBox]])
-
-    checkAll(s"$context.Apply[CaseClassWOption]", applyTests[CaseClassWOption].apply[Int, String, Long])
-    checkAll(s"$context.Apply[OptList]", applyTests[OptList].apply[Int, String, Long])
+    checkAll(s"$context.Apply[CaseClassWOption]", applyTests[CaseClassWOption])
+    checkAll(s"$context.Apply[OptList]", applyTests[OptList])
 //    checkAll(s"$context.Apply[AndInt]", applyTests[AndInt].apply[Int, String, Long])
-    checkAll(s"$context.Apply[Interleaved]", applyTests[Interleaved].apply[Int, String, Long])
-    checkAll(s"$context.Apply[ListBox]", applyTests[ListBox].apply[Int, String, Long])
+    checkAll(s"$context.Apply[Interleaved]", applyTests[Interleaved])
+    checkAll(s"$context.Apply[ListBox]", applyTests[ListBox])
     checkAll(s"$context.Apply is Serializable", SerializableTests.serializable(summonInline[Apply[Interleaved]]))
   }
 
```

But I'm getting even weirder errors. Arbitrary generators cannot be found:
```
[info] compiling 32 Scala sources to D:\MojeProgramy\forki\kittens\core\.jvm\target\scala-3.1.0\test-classes ...
[error] -- Error: D:\MojeProgramy\forki\kittens\core\src\test\scala-3\cats\derived\ApplySuite.scala:33:49
[error] 33 |    ApplyTests[F](apply).apply[Int, String, Long]
[error]    |                                                 ^
[error]    |no implicit argument of type org.scalacheck.Arbitrary[F[Int]] was found for parameter ArbFA of method apply in trait ApplyTests
[error]    |
[error]    |where:    F is a type in method applyTests with bounds <: [_] =>> Any
[error]    |.
[error]    |I found:
[error]    |
[error]    |    org.scalacheck.Arbitrary.arbContainer[F, Int](org.scalacheck.Arbitrary.arbInt,
[error]    |      org.scalacheck.util.Buildable.buildableFactory[Int, F[Int]](
[error]    |        /* missing */summon[collection.Factory[Int, F[Int]]]
[error]    |      )
[error]    |    , ???)
[error]    |
[error]    |But no implicit values were found that match type collection.Factory[Int, F[Int]].
[error] one error found
[error] one error found
[error] (coreJVM / Test / compileIncremental) Compilation failed
```

I don't get it all - maybe implicits in inline methods works to some certain depth?

File: core/src/test/scala-3/cats/derived/ApplySuite.scala (hunk 1 above)
Comment (joroKr21) [reply core/src/test/scala-3/cats/derived/ApplySuite.scala]: Oh ok if doesn't work - forget it 😄 
It's just a test anyway.

Comment (joroKr21) [reply core/src/test/scala-3/cats/derived/ApplySuite.scala]: Oh wait - so it did work 👍

Comment (andrzejressel) [reply core/src/test/scala-3/cats/derived/ApplySuite.scala]: No, no - it works fine. But I feel there is something wrong with how it works - I guess I'll have to do more digging with newer Scalas. I for some weird reason have to declare new implicits when there was not need to before.

**Review by joroKr21 — APPROVED** [review]
Nice 👍

**Pull request merged**